File: lintcode/minimum-depth-of-binary-tree.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/minimum-depth-of-binary-tree/

"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def minDepth(self, root):
        # write your code here
        if not root:
            return 0

        if root.left and root.right:
            return min(1+self.minDepth(root.left), 1+self.minDepth(root.right))
        elif root.left and not root.right:
            return 1+self.minDepth(root.left)
        elif not root.left and root.right:
            return 1+self.minDepth(root.right)
        else:
            return 1

if __name__ == '__main__':
    root = TreeNode(1)
    root.right = TreeNode(3)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)
    root.right.left.left = TreeNode(6)
    s = Solution()
    print s.minDepth(root)

File: leetcode/minimum-depth-of-binary-tree.cc
// Question: https://leetcode.com/problems/minimum-depth-of-binary-tree/

#include "headers.h"

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root) {
            return 0;
        }

        return helper(root);
    }

private:
    int helper(TreeNode* root) {
        if (!root) {
            return INT_MAX;
        }

        int left = helper(root->left);
        int right = helper(root->right);
        if (left == INT_MAX && right == INT_MAX) {
            return 1;
        }
        return min(left, right) + 1;
    }
};

int main(int argc, char** argv) {
    Solution s;
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->left->left = new TreeNode(3);
    root->right = new TreeNode(4);
    cout << s.minDepth(root) << endl;
    cout << s.minDepth(NULL) << endl;
    return 0;
}
